File: src/tech/v3/dataset/modelling.clj
(ns tech.v3.dataset.modelling
  "Methods related specifically to machine learning such as setting the inference
  target.  This file integrates tightly with tech.v3.dataset.categorical which provides
  categorical -> number and one-hot transformation pathways"
  (:require [tech.v3.datatype :as dtype]
            [tech.v3.datatype.errors :as errors]
            [tech.v3.datatype.casting :as casting]
            [tech.v3.datatype.argops :as argops]
            [tech.v3.dataset.base :as ds-base]
            [tech.v3.dataset.readers :as ds-readers]
            [tech.v3.dataset.categorical :as categorical]
            [tech.v3.dataset.column :as ds-col]
            [clojure.set :as set])
  (:import [tech.v3.datatype ObjectReader]
           [java.util List]))


(set! *warn-on-reflection* true)


(defn inference-column?
  [col]
  (:inference-target? (meta col)))


(defn set-inference-target
  "Set the inference target on the column.  This sets the :column-type member
  of the column metadata to :inference-target?."
  [dataset target-name-or-target-name-seq]
  (let [colnames (if (sequential? target-name-or-target-name-seq)
                   target-name-or-target-name-seq
                   [target-name-or-target-name-seq])]
    (ds-base/update-columns dataset colnames
                            #(vary-meta % assoc :inference-target? true))))


(defn inference-target-column-names
  "Return the names of the columns that are inference targets."
  [ds]
  (->> (map meta (vals ds))
       (filter :inference-target?)
       (map :name)
       (seq)))


(defn inference-target-label-map
  [dataset & [label-columns]]
  (let [label-columns (or label-columns (inference-target-column-names dataset))]
    (errors/when-not-errorf
     (= 1 (count label-columns))
     "Multiple or zero label columns found: %s" label-columns)
    (-> (ds-base/column dataset (first label-columns))
        (meta)
        (get-in [:categorical-map :lookup-table]))))


(defn inference-target-label-inverse-map
  "Given options generated during ETL operations and annotated with :label-columns
  sequence container 1 label column, generate a reverse map that maps from a dataset
  value back to the label that generated that value."
  [dataset & [label-columns]]
  (set/map-invert (inference-target-label-map dataset label-columns)))



(defn dataset->categorical-xforms
  "Given a dataset, return a map of column-name->xform information."
  [ds]
  (->> (concat (categorical/dataset->categorical-maps ds)
               (categorical/dataset->one-hot-maps ds))
       (map (juxt :src-column identity))
       (into {})))


(defn num-inference-classes
  "Given a dataset and correctly built options from pipeline operations,
  return the number of classes used for the label.  Error if not classification
  dataset."
  ^long [dataset]
  (count (inference-target-label-map dataset)))


(defn feature-ecount
  "Number of feature columns.  Feature columns are columns that are not
  inference targets."
  ^long [dataset]
  (count (remove #(= :inference-target? (meta %))
                 (vals dataset))))


(defn model-type
  "Check the label column after dataset processing.
  Return either
  :regression
  :classification"
  [dataset & [column-name-seq]]
  (->> (or column-name-seq
           (inference-target-column-names dataset))
       (ds-base/select-columns dataset)
       (vals)
       (map (fn [column]
              (let [colmeta (meta column)]
                [(:name colmeta) (if (:categorical? colmeta)
                                   :classification
                                   :regression)])))
       (into {})))


(defn column-values->categorical
  "Given a column encoded via either string->number or one-hot, reverse
  map to the a sequence of the original string column values.
  In the case of one-hot mappings, src-column must be the original
  column name before the one-hot map"
  [dataset src-column]
  (if-let [cmap (->> (categorical/dataset->categorical-maps dataset)
                     (filter #(= src-column (:src-column %)))
                     (first))]
    (-> (categorical/invert-categorical-map dataset cmap)
        (ds-base/column src-column))
    (if-let [one-hot-map (->> (categorical/dataset->one-hot-maps dataset)
                              (filter #(= src-column (:src-column %)))
                              (first))]
      (-> (categorical/invert-one-hot-map dataset one-hot-map)
          (ds-base/column src-column))
      (errors/throwf
       "Column %s does not appear to have either a categorical or one hot map"
       src-column))))


(defn- dataset-indexes
  [^long n-rows {:keys [randomize-dataset?]
                 :or {randomize-dataset? true}
                 :as options}]
  (if randomize-dataset?
    (argops/argshuffle n-rows options)
    (range n-rows)))


(defn k-fold-datasets
  "Given 1 dataset, prepary K datasets using the k-fold algorithm.
  Randomize dataset defaults to true which will realize the entire dataset
  so use with care if you have large datasets.

  Returns a sequence of {:test-ds :train-ds}

  Options:

  * `:randomize-dataset?` - When true, shuffle the dataset.  In that case 'seed' may be
     provided.  Defaults to true.
  * `:seed` -  when `:randomize-dataset?` is true then this can either be an
     implementation of java.util.Random or an integer seed which will be used to
     construct java.util.Random."
  ([dataset k options]
   (let [n-rows (ds-base/row-count dataset)
         indexes (dataset-indexes n-rows options)
         indexes-per-fold (quot n-rows k)
         leftover (rem n-rows k)
         folds (mapv (fn [^long idx]
                       (let [start-idx (-> (* idx indexes-per-fold)
                                           (+ (if (< idx leftover)
                                                idx
                                                0)))
                             glen (+ indexes-per-fold
                                     (if (< idx leftover)
                                       1
                                       0))]
                         (dtype/sub-buffer indexes start-idx glen)))
                     (range k))]
     (for [i (range k)]
       {:test-ds (ds-base/select-rows dataset (nth folds i))
        :train-ds (ds-base/select-rows dataset
                                       (->> (keep-indexed #(if (not= %1 i) %2) folds)
                                            (apply concat)))})))
  ([dataset k]
   (k-fold-datasets dataset k {})))


(defn train-test-split
  "Probabilistically split the dataset returning a map of `{:train-ds :test-ds}`.

  Options:

  * `:randomize-dataset?` - When true, shuffle the dataset.  In that case 'seed' may be
     provided.  Defaults to true.
  * `:seed` -  when `:randomize-dataset?` is true then this can either be an
     implementation of java.util.Random or an integer seed which will be used to
     construct java.util.Random.
  * `:train-fraction` - Fraction of the dataset to use as training set.  Defaults to
     0.7."
  ([dataset {:keys [train-fraction]
             :or {train-fraction 0.7}
             :as options}]
   (let [[_n-cols n-rows] (dtype/shape dataset)
         indexes (dataset-indexes n-rows options)
         n-elems (long n-rows)
         n-training (long (Math/round (* n-elems (double train-fraction))))]
     {:train-ds (ds-base/select-rows dataset (take n-training indexes))
      :test-ds (ds-base/select-rows dataset (drop n-training indexes))}))
  ([dataset]
   (train-test-split dataset {})))


(defn inference-target-ds
  "Given a dataset return reverse-mapped inference target columns or nil
  in the case where there are no inference targets."
  [dataset]
  (when-let [target-cols (inference-target-column-names dataset)]
    (-> (ds-base/select-columns dataset target-cols)
        (categorical/reverse-map-categorical-xforms))))


(defn labels
  "Return the labels.  The labels sequence is the reverse mapped inference
  column.  This returns a single column of data or errors out."
  [dataset]
  (let [
        rev-mapped (inference-target-ds dataset)]
    (errors/when-not-errorf
        (== 1 (ds-base/column-count rev-mapped))
      "Incorrect number of columns (%d) in dataset for labels transformation"
      (ds-base/column-count rev-mapped))
    rev-mapped))


(defn probability-distributions->label-column
  "Given a dataset that has columns in which the column names describe labels and the
  rows describe a probability distribution, create a label column by taking the max
  value in each row and assign column that row value."
  [prob-ds dst-colname]
  (let [^List cnames (vec (ds-base/column-names prob-ds))
        idx-col (->> (ds-readers/value-reader prob-ds)
                     (dtype/emap
                      (fn [valvec]
                        (let [rdr (dtype/->reader valvec :float64)]
                          ;;scan for missing and throw
                          (dotimes [idx (.lsize rdr)]
                            (errors/when-not-errorf
                             (Double/isFinite (.readDouble rdr idx))
                             "Nan/infinite values not allowed in probability distributions"))
                          (argops/argmax rdr)))
                      :float64)
                     (dtype/clone))
        cat-map (categorical/create-categorical-map
                 (->> (map-indexed vector cnames)
                      (into {})
                      (set/map-invert))
                 dst-colname
                 :float64)
        retval
        (assoc prob-ds dst-colname
               (ds-col/new-column
                dst-colname
                idx-col
                {:categorical? true
                 :categorical-map cat-map}
                (ds-base/missing prob-ds)))]
    retval))

File: java_public_api/tech/v3/dataset/Modelling.java
package tech.v3.dataset;


import static tech.v3.Clj.*;
import clojure.lang.IFn;
import clojure.lang.Keyword;
import java.util.Map;


/**
 * Functions related to training and evaluating ML models.  The functions are grouped into
 * a few groups.
 *
 * For the purpose of this system, categorical data means a column of data that is not numeric.
 * it could be strings, keywords, or arbitrary objects.
 *
 * PCA *for now* requires either opencpp's blas library or smile's MKL bindings.  Expect this
 * requirement to change to solely requiring neanderthal in the near future.
 *
 * Minimal example extra dependencies for PCA:
 *
 *```console
 * [org.bytedeco/openblas "0.3.10-1.5.4"]
 * [org.bytedeco/openblas-platform "0.3.10-1.5.4"]
 *```
 *
 * It is also important to note that you can serialize the fit results to nippy automatically
 * as included in dtype-next are extensions to nippy that work with tensors.
 */
public class Modelling {
  private Modelling(){}

  static final IFn fitCatFn = requiringResolve("tech.v3.dataset.categorical", "fit-categorical-map");
  static final IFn transCatFn = requiringResolve("tech.v3.dataset.categorical", "transform-categorical-map");
  static final IFn invCatFn = requiringResolve("tech.v3.dataset.categorical", "invert-categorical-map");
  static final IFn fitOneHotFn = requiringResolve("tech.v3.dataset.categorical", "fit-one-hot");
  static final IFn transOneHotFn = requiringResolve("tech.v3.dataset.categorical", "transform-one-hot");
  static final IFn invOneHotFn = requiringResolve("tech.v3.dataset.categorical", "invert-one-hot");

  static final IFn corrTableFn = requiringResolve("tech.v3.dataset.math", "correlation-table");
  static final IFn fillRangeReplaceFn = requiringResolve("tech.v3.dataset.math", "fill-range-replace");
  static final IFn fitPCAFn = requiringResolve("tech.v3.dataset.math", "fit-pca");
  static final IFn fitStdScaleFn = requiringResolve("tech.v3.dataset.math", "fit-std-scale");
  static final IFn fitMinMaxFn = requiringResolve("tech.v3.dataset.math", "fit-minmax");
  static final IFn transPCAFn = requiringResolve("tech.v3.dataset.math", "transform-pca");
  static final IFn transStdScaleFn = requiringResolve("tech.v3.dataset.math", "transform-std-scale");
  static final IFn transMinMaxFn = requiringResolve("tech.v3.dataset.math", "transform-minmax");
  static final IFn interpolateLOESSFn = requiringResolve("tech.v3.dataset.math", "interpolate-loess");

  static final IFn kFoldFn = requiringResolve("tech.v3.dataset.modelling", "k-fold-datasets");
  static final IFn trainTestFn = requiringResolve("tech.v3.dataset.modelling", "train-test-split");
  static final IFn setInfTargetFn = requiringResolve("tech.v3.dataset.modelling", "set-inference-target");
  static final IFn labelsFn = requiringResolve("tech.v3.dataset.modelling", "labels");
  static final IFn probDistToLabel = requiringResolve("tech.v3.dataset.modelling", "probability-distributions->label-column");
  static final IFn infTargetLabelMap = requiringResolve("tech.v3.dataset.modelling", "inference-target-label-map");


  /**
   * Fit an object->integer transform that takes each value and assigned an integer to it.  The
   * returned value can be used in transformCategorical to transform the dataset.
   *
   * Options:
   *
   * * `:table-args` - Either a sequence of vectors [col-val, idx] or a sorted sequence
   *   of column values where integers will be assigned as per the sorted sequence.  Any values
   *   found outside the the specified values will be auto-mapped to the next largest integer.
   * * `:res-dtype` - Datatype of result column.  Defaults to `:float64`.
   */
  public static Map fitCategorical(Object ds, Object cname, Object options) {
    return (Map)fitCatFn.invoke(ds, cname, options);
  }
  /**
   * Fit an object->integer transformation.  Integers will be assigned in random order.  For
   * more control over the transform see the 3-arity version of the function.
   */
  public static Map fitCategorical(Object ds, Object cname) {
    return (Map)fitCatFn.invoke(ds, cname);
  }
  /**
   * Apply an object->integer transformation with data obtained from fitCategorical.
   */
  public static Map transformCategorical(Object ds, Object catFitData) {
    return (Map)transCatFn.invoke(ds, catFitData);
  }
  /**
   * Reverse a previously transformed categorical mapping.
   */
  public static Map invertCategorical(Object ds, Object catFitData) {
    return (Map)invCatFn.invoke(ds, catFitData);
  }
  /**
   * Fit a transformation from a single column of categorical values to a `one-hot` encoded
   * group of columns.
   * .
   *
   * Options:
   *
   * * `:table-args` - Either a sequence of vectors [col-val, idx] or a sorted sequence
   *   of column values where integers will be assigned as per the sorted sequence.  Any values
   *   found outside the the specified values will be auto-mapped to the next largest integer.
   * * `:res-dtype` - Datatype of result column.  Defaults to `:float64`.
   *
   */
  public static Map fitOneHot(Object ds, Object cname, Object options) {
    return (Map)fitOneHotFn.invoke(ds, cname, options);
  }
  /**
   * Fit a mapping from a categorical column to a group of one-hot encoded columns.
   */
  public static Map fitOneHot(Object ds, Object cname) {
    return (Map)fitOneHotFn.invoke(ds, cname);
  }
  /**
   * Transform a dataset using a fitted one-hot mapping.
   */
  public static Map transformOneHot(Object ds, Object fitData) {
    return (Map)transOneHotFn.invoke(ds, fitData);
  }
  /**
   * Reverse a previously transformed one-hot mapping.
   */
  public static Map invertOneHot(Object ds, Object fitData) {
    return (Map)invOneHotFn.invoke(ds, fitData);
  }
  /**
   * Return a map of column to inversely sorted from greatest to least sequence of tuples of
   * column name, coefficient.
   *
   * Options:
   *
   * * `:correlation-type` One of `:pearson`, `:spearman`, or `:kendall`.  Defaults to
   *   `:pearson`.
   */
  public static Map correlationTable(Object ds, Object options) {
    return (Map)corrTableFn.invoke(ds, options);
  }
  /**
   * Return a map of column to inversely sorted from greatest to least sequence of tuples of
   * column name, pearson correlation coefficient.
   */
  public static Map correlationTable(Object ds) {
    return (Map)corrTableFn.invoke(ds);
  }
  /**
   * Expand a dataset ensuring that the difference between two successive values is less than
   * `max-span`.
   *
   * @param maxSpan The minimal span value.  For datetime types this is interpreted in
   * millisecond or epoch-millisecond space.
   * @param missingStrategy Same missing strategy types from `TMD.replaceMissing`.
   */
  public static Map fillRangeReplace(Object ds, Object cname, double maxSpan, Object missingStrategy) {
    Keyword strat;
    Object value = null;
    if (isVector(missingStrategy)) {
	strat = (Keyword)call(missingStrategy, 0);
	value = call(missingStrategy, 1);
      } else {
      strat = (Keyword)missingStrategy;
    }
    return (Map) fillRangeReplaceFn.invoke(ds, cname, maxSpan, strat, value);
  }
  /**
   * Fit a PCA transformation on a dataset.
   *
   * @return map of `{:means, :eigenvalues, :eigenvectors}`.
   *
   * Options:
   *
   * * `:method` - either `:svd` or `cov`.  Use either SVD transformation or covariance-matrix
   *   base PCA.  `:cov` method is somewhat slower but returns accurate variances.
   * * `:variance-amount` - Keep columns until variance is just less than variance-amount.
   *   Defaults to 0.95.
   * * `:n-components` - Return a fixed number of components.  Overrides `:variance-amount`
   *   an returns a fixed number of components.
   * * `:covariance-bias` - When using `:cov` divide by `n-rows` if true and `n-rows - 1` if
   *   false.  Defaults to false.
   */
  public static Object fitPCA(Object ds, Object options) {
    return fitPCAFn.invoke(ds, options);
  }
  /**
   * Fit a PCA transformation onto a dataset keeping 95% of the variance.  See documentation
   * for 2-arity form.
   */
  public static Object fitPCA(Object ds) {
    return fitPCAFn.invoke(ds);
  }
  /**
   * Transform a dataset by the PCA fit data.
   */
  public static Map transformPCA(Object ds, Object fitData) {
    return (Map)transPCAFn.invoke(ds, fitData);
  }
  /**
   * Calculate per-column mean, stddev.
   *
   * Options:
   *
   * * `:mean?` - Produce per-column means.  Defaults to true.
   * * `:stddev?` - Produce per-column standard deviation.  Defaults to true.
   */
  public static Object fitStdScale(Object ds) {
    return fitStdScaleFn.invoke(ds);
  }
  /**
   * Transform dataset to mean of zero and a standard deviation of 1.
   */
  public static Map transformStdScale(Object ds, Object fitData) {
    return (Map)transStdScaleFn.invoke(ds, fitData);
  }
  /**
   * Fit a bias and scale the dataset that transforms each colum to a target min-max
   * value.
   *
   * Options:
   *
   * * `:min` - Target minimum value.  Defaults it -0.5.
   * * `:max` - Target maximum value.  Defaults to 0.5.
   */
  public static Object fitMinMax(Object ds, Object options) {
    return fitMinMaxFn.invoke(ds, options);
  }
  /**
   * Fit a minmax transformation that will transform each column to a minimum of -0.5 and
   * a maximum of 0.5.
   */
  public static Object fitMinMax(Object ds) {
    return fitMinMaxFn.invoke(ds);
  }
  /**
   * Transform a dataset using a previously fit minimax transformation.
   */
  public static Map transformMinMax(Object ds, Object fitData) {
    return (Map)transMinMaxFn.invoke(ds, fitData);
  }
  /**
   * Map a LOESS-interpolation transformation onto a dataset.  This can be used
   * to, among other things, smooth out a column before graphing.  For the meaning
   * of the options, see documentation on the
   * org.apache.commons.math3.analysis.interpolationLoessInterpolator.
   *
   * Option defaults have been chosen to map somewhat closely to the R defaults.
   *
   * Options:
   *
   * * `:bandwidth` - Defaults to 0.75.
   * * `:iterations` - Defaults to 4.
   * * `:accuracy` - Defaults to LoessInterpolator/DEFAULT_ACCURACY.
   * * `:result-name` - Result column name.  Defaults to `yColname.toString +  "-loess"`.
   */
  public static Map interpolateLOESS(Object ds, Object xColname, Object yColname, Object options) {
    return (Map)interpolateLOESSFn.invoke(ds, xColname, yColname, options);
  }
  /**
   * Perform a LOESS interpolation using the default parameters.  For options see 4-arity
   * form of function.
   */
  public static Map interpolateLOESS(Object ds, Object xColname, Object yColname) {
    return (Map)interpolateLOESSFn.invoke(ds, xColname, yColname);
  }
  /**
   * Produce 2*k datasets from 1 dataset using k-fold algorithm.
   * Returns a k maps of the form `{:test-ds :train-ds}.
   *
   * Options:
   *
   * * `:randomize-dataset?` - When true, shuffle dataset.  Defaults to true.
   * * `:seed` - When randomizing dataset, seed may be either an integer or an implementation
   *   of `java.util.Random`.
   */
  public static Iterable kFold(Object ds, long k, Object options) {
    return (Iterable)kFoldFn.invoke(ds, k, options);
  }
  /**
   * Return k maps of the form `{:test-ds :train-ds}`.  For options see 3-arity form.
   */
  public static Iterable kFold(Object ds, long k) {
    return (Iterable)kFoldFn.invoke(ds, k);
  }
  /**
   * Split the dataset returning a map of `{:train-ds :test-ds}`.
   *
   * Options:
   *
   * * `:randomize-dataset?` - Defaults to true.
   * * `:seed` - When provided must be an integer or an implementation `java.util.Random`.
   * * `:train-fraction` - Fraction of dataset to use as training set.  Defaults to 0.7.
   */
  public static Map trainTestSplit(Object ds, Object options) {
    return (Map)trainTestFn.invoke(ds, options);
  }
  /**
   * Randomize then split dataset using 70% of the data for training and the rest for testing.
   */
  public static Map trainTestSplit(Object ds) {
    return (Map)trainTestFn.invoke(ds);
  }
  /**
   * Set a column in the dataset as the inference target.  This information is stored in the
   * column metadata.  This function is short form for:
   *
   *```java
   *  Object col = column(ds, cname);
   *  return assoc(ds, cname, varyMeta(col, assocFn, kw("inference-target?"), true));
   *```
   */
  public static Map setInferenceTarget(Object ds, Object cname) {
    return (Map)setInfTargetFn.invoke(ds, cname);
  }
  /**
   * Find the inference column.  If column was the result of a categorical mapping, reverse
   * that mapping.  Return data in a form that can be efficiently converted to a Buffer.
   */
  public static Object labels(Object ds) {
    return labelsFn.invoke(ds);
  }
  /**
   * Given a dataset where the column names are labels and the each row is a probabilitly
   * distribution across the labels, produce a Buffer of labels taking the highest probability
   * for each row to choose the label.
   */
  public static Object probabilityDistributionToLabels(Object ds) {
    return probDistToLabel.invoke(ds);
  }
  /**
   * Return a map of val->idx for the inference target.
   */
  public static Map inferenceTargetLabelMap(Object ds) {
    return (Map)infTargetLabelMap.invoke(ds);
  }
}
